R/if2.R

```diff
@@ -0,0 +1,279 @@
+##' Run an IF2 inference. This is experimental and subject to change. Use at
+##' your own risk.
+##'
+##' @title Run iterated filtering (IF2 algorithm)
+##'
+##' @description Create an IF2 object for running
+##'   and interacting with an IF2 inference.
+##'
+##' @export
+##' @importFrom R6 R6Class
+##' @examples
+##' A basic SIR model used in the particle filter example
+##' dat <- example_sir()
+##'
+##' # Range and initial values for model parameters
+##' pars <- if2_parameters$new(
+##'           list(if2_parameter("beta", 0.15, min = 0, max = 1),
+##'                if2_parameter("gamma", 0.05, min = 0, max = 1)))
+##'
+##' # Set up of IF2 algorithm
+##' iterations <- 100
+##' cooling_target <- 0.5
+##' n_par_sets <- 300
+##' control <- if2_control(pars_sd = list("beta" = 0.02, "gamma" = 0.02),
+##'                         iterations = iterations,
+##'                         n_par_sets = n_par_sets,
+##'                         cooling_target = cooling_target,
+##'                         progress = FALSE)
+##'
+##' Set up, and run IF2
+##' filter <- if2$new(pars, dat$data, dat$model, dat$compare, NULL,
+##'                   dat$index, control)
+##' filter$run()
+##'
+##' # Plot results
+##' filter$plot("ll")
+##' filter$plot("beta")
+##' filter$plot("gamma")
+##'
+##' # Get log-likelihood estimates from running a particle filter at
+##' # each final parameter estimate
+##' n_particles <- 100
+##' ll_samples <- filter$sample(n_particles)
+if2 <- R6::R6Class(
```

Comment:
This is quite a different interface to that presented by pmcmc and I'd like to harmonise if possible. There, we have the functional interface where calling the function `pmcmc` returns a plain list of class `mcstate_pmcmc` which has pars, likelihoods etc. On the way it uses a semi-exported object pmcmc_state but that's an implementation detail.

So here, the `if2` function should use the object much as you have it, but return all the requested information as a list. Unless there's some reason that's not possible, in which case please let me know.